File: backend/src/test/java/com/backend/api/user/controller/MyPageControllerTest.java
package com.backend.api.user.controller;

import com.backend.api.user.dto.response.UserMyPageResponse;
import com.backend.api.user.service.UserMyPageService;
import com.backend.domain.user.entity.Role;
import com.backend.domain.user.entity.User;
import com.backend.domain.user.repository.UserRepository;
import com.backend.global.Rq.Rq;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.transaction.Transactional;
import org.junit.jupiter.api.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.util.ReflectionTestUtils;
import org.springframework.test.web.servlet.MockMvc;
import org.junit.jupiter.api.*;
import org.springframework.http.MediaType;

import java.util.Map;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc(addFilters = false)
@Transactional
@ActiveProfiles("test")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class MyPageControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private UserRepository userRepository;

    @MockBean
    private UserMyPageService userMyPageService;

    @MockBean
    private Rq rq;

    private User user;

    @BeforeEach
    void setUp() {

        user = User.builder()
                .email("user@test.com")
                .password("user1234!")
                .name("일반유저")
                .nickname("user")
                .age(25)
                .github("github.com/user")
                .image(null)
                .role(Role.USER)
                .build();

        ReflectionTestUtils.setField(user, "id", 1L);
        when(rq.getUser()).thenReturn(user);
    }

    @Nested
    @DisplayName("사용자 전체 API")
    class test1{
        @Test
        @DisplayName("로그인 상태")
        void success() throws Exception {
            UserMyPageResponse response = UserMyPageResponse.builder()
                    .userId(user.getId())
                    .email(user.getEmail())
                    .name(user.getName())
                    .nickname(user.getNickname())
                    .age(user.getAge())
                    .github(user.getGithub())
                    .image(user.getImage())
                    .build();

            when(userMyPageService.getInformation(user.getId())).thenReturn(response);

            mockMvc.perform(get("/api/v1/users/me"))
                    .andExpect(status().isOk())
                    .andExpect(jsonPath("$.status").value("OK"))
                    .andExpect(jsonPath("$.data.email").value("user@test.com"))
                    .andExpect(jsonPath("$.data.name").value("일반유저"))
                    .andDo(print());
        }

        @Test
        @DisplayName("비로그인 상태")
        void fail1() throws Exception {

            when(rq.getUser()).thenReturn(null);

            mockMvc.perform(get("/api/v1/users/me"))
                    .andExpect(status().isUnauthorized())
                    .andExpect(jsonPath("$.status").value("UNAUTHORIZED"))
                    .andExpect(jsonPath("$.message").value("로그인된 사용자가 없습니다."))
                    .andDo(print());
        }
    }

    @Nested
    @DisplayName("개인정보 비밀번호 API")
    class t2 {

        @Test
        @DisplayName("성공 - 비밀번호 일치")
        void success() throws Exception {
            Map<String, String> requestBody = Map.of("password", "user1234!");
            when(userMyPageService.verifyPassword(user.getId(), "user1234!")).thenReturn(true);

            mockMvc.perform(post("/api/v1/users/verifyPassword")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(requestBody)))
                    .andExpect(status().isOk())
                    .andExpect(jsonPath("$.status").value("OK"))
                    .andExpect(jsonPath("$.message").value("비밀번호가 확인되었습니다."))
                    .andExpect(jsonPath("$.data").value(true))
                    .andDo(print());
        }

        @Test
        @DisplayName("실패 - 비밀번호 불일치")
        void fail1() throws Exception {
            Map<String, String> requestBody = Map.of("password", "wrongPassword");
            when(userMyPageService.verifyPassword(user.getId(), "wrongPassword")).thenReturn(false);

            mockMvc.perform(post("/api/v1/users/verifyPassword")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(requestBody)))
                    .andExpect(status().isUnauthorized())
                    .andExpect(jsonPath("$.status").value("UNAUTHORIZED"))
                    .andExpect(jsonPath("$.message").value("비밀번호가 일치하지 않습니다."))
                    .andDo(print());
        }
    }

    @Nested
    @DisplayName("마이페이지 개인 정보 수정 API")
    class t3 {

        @Test
        @DisplayName("성공 - 사용자 정보 수정")
        void success() throws Exception {
            Map<String, Object> modify = Map.of(
                    "email", "새 이메일",
                    "password", "새 비밀번호",
                    "name", "새 이름",
                    "nickname", "새 닉네임",
                    "age", 27,
                    "github", "newGithub",
                    "image", "newImage"
            );

            UserMyPageResponse response = UserMyPageResponse.builder()
                    .userId(user.getId())
                    .email("새 이메일")
                    .name("새 이름")
                    .nickname("새 닉네임")
                    .age(27)
                    .github("newGithub")
                    .image("newImage")
                    .build();

            when(userMyPageService.modifyUser(eq(user.getId()), any(UserMyPageResponse.UserModify.class)))
                    .thenReturn(response);

            mockMvc.perform(put("/api/v1/users/me")
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(objectMapper.writeValueAsString(modify)))
                    .andExpect(status().isOk())
                    .andExpect(jsonPath("$.status").value("OK"))
                    .andExpect(jsonPath("$.message").value("개인 정보 수정이 완료되었습니다."))
                    .andExpect(jsonPath("$.data.nickname").value("새 닉네임"))
                    .andDo(print());
        }

    }

}

File: backend/src/test/java/com/backend/api/user/controller/MyPageControllerTest.kt
package com.backend.api.user.controller

import com.backend.api.user.dto.response.UserMyPageResponse
import com.backend.api.user.dto.request.MyPageRequest.UserModify
import com.backend.api.user.service.UserMyPageService
import com.backend.domain.user.entity.Role
import com.backend.domain.user.entity.User
import com.backend.domain.user.repository.UserRepository
import com.backend.global.Rq.Rq
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.transaction.Transactional
import org.junit.jupiter.api.*
import org.mockito.Mockito
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.MediaType
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.TestConstructor
import org.springframework.test.util.ReflectionTestUtils
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultHandlers
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

import org.mockito.kotlin.whenever
import org.mockito.kotlin.any
import org.mockito.kotlin.eq

@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)

@SpringBootTest
@AutoConfigureMockMvc(addFilters = false)
@Transactional
@ActiveProfiles("test")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class MyPageControllerTest(
    private val mockMvc: MockMvc,
    private val objectMapper: ObjectMapper,
    private val userRepository: UserRepository) {

    @MockBean
    lateinit var userMyPageService: UserMyPageService

    @MockBean
    lateinit var rq: Rq

    private lateinit var user: User

    @BeforeEach
    fun setUp() {
        user = User(
            email = "user@test.com",
            password = "user1234!",
            name = "일반유저",
            nickname = "user",
            age = 25,
            github = "github.com/user",
            image = null,
            role = Role.USER
        )
        ReflectionTestUtils.setField(user, "id", 1L)
        Mockito.`when`(rq.getUser()).thenReturn(user)
    }

    @Nested
    @DisplayName("사용자 전체 API")
    inner class Test1 {
        @Test
        @DisplayName("로그인 상태")
        fun success() {
            val response = UserMyPageResponse.fromEntity(user)
            Mockito.`when`(userMyPageService.getInformation(user.id)).thenReturn(response)

            mockMvc.perform(MockMvcRequestBuilders.get("/api/v1/users/me"))
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.jsonPath("$.status").value("OK"))
                .andExpect(MockMvcResultMatchers.jsonPath("$.data.email").value("user@test.com"))
                .andExpect(MockMvcResultMatchers.jsonPath("$.data.name").value("일반유저"))
                .andDo(MockMvcResultHandlers.print())
        }

        @Test
        @DisplayName("비로그인 상태")
        fun fail1() {
            Mockito.`when`(rq.getUser()).thenThrow(com.backend.api.user.controller.testsupport.UnauthenticatedException())

            mockMvc.perform(MockMvcRequestBuilders.get("/api/v1/users/me"))
                .andExpect(MockMvcResultMatchers.status().isUnauthorized)
                .andExpect(MockMvcResultMatchers.jsonPath("$.status").value("UNAUTHORIZED"))
                .andExpect(MockMvcResultMatchers.jsonPath("$.message").value("로그인된 사용자가 없습니다."))
                .andDo(MockMvcResultHandlers.print())
        }
    }

    @Nested
    @DisplayName("개인정보 비밀번호 API")
    inner class Test2 {
        @Test
        @DisplayName("성공 - 비밀번호 일치")
        fun success() {
            val requestBody = mapOf("password" to "user1234!")
            Mockito.`when`(userMyPageService.verifyPassword(user.id, "user1234!")).thenReturn(true)

            mockMvc.perform(
                MockMvcRequestBuilders.post("/api/v1/users/verifyPassword")
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(requestBody))
            )
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.jsonPath("$.status").value("OK"))
                .andExpect(MockMvcResultMatchers.jsonPath("$.message").value("비밀번호가 확인되었습니다."))
                .andExpect(MockMvcResultMatchers.jsonPath("$.data").value(true))
                .andDo(MockMvcResultHandlers.print())
        }

        @Test
        @DisplayName("실패 - 비밀번호 불일치")
        fun fail1() {
            val requestBody = mapOf("password" to "wrongPassword")
            Mockito.`when`(userMyPageService.verifyPassword(user.id, "wrongPassword")).thenReturn(false)

            mockMvc.perform(
                MockMvcRequestBuilders.post("/api/v1/users/verifyPassword")
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(requestBody))
            )
                .andExpect(MockMvcResultMatchers.status().isUnauthorized)
                .andExpect(MockMvcResultMatchers.jsonPath("$.status").value("UNAUTHORIZED"))
                .andExpect(MockMvcResultMatchers.jsonPath("$.message").value("비밀번호가 일치하지 않습니다."))
                .andDo(MockMvcResultHandlers.print())
        }
    }

    @Nested
    @DisplayName("마이페이지 개인 정보 수정 API")
    inner class Test3 {
        @Test
        @DisplayName("성공 - 사용자 정보 수정")
        fun success() {
            val modify = mapOf(
                "email" to "새 이메일",
                "password" to "새 비밀번호",
                "name" to "새 이름",
                "nickname" to "새 닉네임",
                "age" to 27,
                "github" to "newGithub",
                "image" to "newImage"
            )

            val response = UserMyPageResponse.fromEntity(user)

            whenever(
                userMyPageService.modifyUser(
                    eq(user.id),
                    any<UserModify>()
                )
            ).thenReturn(response)

            mockMvc.perform(
                MockMvcRequestBuilders.put("/api/v1/users/me")
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(modify))
            )
                .andExpect(MockMvcResultMatchers.status().isOk)
                .andExpect(MockMvcResultMatchers.jsonPath("$.status").value("OK"))
                .andExpect(MockMvcResultMatchers.jsonPath("$.message").value("개인 정보 수정이 완료되었습니다."))
                .andDo(MockMvcResultHandlers.print())
        }
    }
}